src/vea/vea_free.c

```diff
@@ -562,6 +941,134 @@ need_aging_flush(struct vea_space_info *vsi, uint32_t cur_time, bool force)
 	return true;
 }
 
+struct reclaim_commit_cb_arg {
+	struct vea_space_info	*rca_vsi;
+	struct vea_free_entry	 rca_vfe;
+};
+
+static void
+reclaim_commit_cb(void *data, bool noop)
+{
+	struct reclaim_commit_cb_arg	*rca = data;
+	int				 rc;
+
+	/* Transaction aborted, only need to free callback arg */
+	if (noop)
+		goto free;
+
+	/*
+	 * Aggregated free will be executed on outermost transaction
+	 * commit.
+	 *
+	 * If it fails, the freed space on persistent free tree won't
+	 * be added in in-memory free tree, hence the space won't be
+	 * visible for allocation until the tree sync up on next server
+	 * restart. Such temporary space leak is tolerable, what we must
+	 * avoid is the contrary case: in-memory tree update succeeds
+	 * but persistent tree update fails, which risks data corruption.
+	 */
+	rc = aggregated_free(rca->rca_vsi, &rca->rca_vfe);
+
+	D_CDEBUG(rc, DLOG_ERR, DB_IO, "Aggregated free on vsi:%p rc %d\n",
+		 rca->rca_vsi, rc);
+free:
+	D_FREE(rca);
+}
+
+static int
+reclaim_unused_bitmap(struct vea_space_info *vsi, uint32_t nr_reclaim, uint32_t *nr_reclaimed)
+{
+	int				 i;
+	struct vea_bitmap_entry		*bitmap_entry, *tmp_entry;
+	struct vea_free_bitmap		*vfb;
+	d_iov_t				 key;
+	int				 rc = 0;
+	struct reclaim_commit_cb_arg	*rca;
+	struct umem_instance		*umem = vsi->vsi_umem;
+	int				 nr = 0;
+	uint64_t			 blk_off;
+	uint32_t			 blk_cnt;
+
+	for (i = 0; i < VEA_MAX_BITMAP_CLASS; i++) {
+		d_list_for_each_entry_safe(bitmap_entry, tmp_entry,
+				&vsi->vsi_class.vfc_bitmap_empty[i], vbe_link) {
+			vfb = &bitmap_entry->vbe_bitmap;
+			D_ASSERT(vfb->vfb_class == i + 1);
+			if (!is_bitmap_empty(vfb->vfb_bitmaps, vfb->vfb_bitmap_sz))
```

Comment:
How can a bitmap chunk non-empty when it's in empty list? BTW, I don't quite quite see why an empty chunk shouldn't be on LRU list, I think we just need to track full bitmap chunk in separate list so that we could skip them on allocation.